File: app/javascript/stores/toast.ts
import { defineStore } from 'pinia'
import { useToast } from 'bootstrap-vue-next'

export const useToastStore = defineStore('toast', () => {
  const toast = useToast()

  function show(message, options = {}) {
    toast?.show({
      props: {
        body: message,
        title: options.title || 'Notification',
        variant: options.variant || 'info',
        solid: true
      }
    })
  }

  function success(message, title = 'Success') {
    show(message, { title, variant: 'success' })
  }

  function error(message, title = 'Error') {
    show(message, { title, variant: 'danger' })
  }

  function warning(message, title = 'Warning') {
    show(message, { title, variant: 'warning' })
  }

  function info(message, title = 'Info') {
    show(message, { title, variant: 'info' })
  }

  return {
    show,
    success,
    error,
    warning,
    info
  }
})

File: app/models/toast.rb
# frozen_string_literal: true

# Canonical toast response object. Enforces the {title, message: Array, variant}
# contract at construction time so controllers cannot produce invalid shapes.
class Toast
  attr_reader :title, :message, :variant

  VALID_VARIANTS = %w[danger warning success info].freeze

  def initialize(title:, message:, variant: 'danger')
    @title = title
    @message = Array(message).freeze
    @variant = variant
  end

  def as_json(_options = nil)
    { 'title' => @title, 'message' => @message, 'variant' => @variant }
  end

  def to_json(*)
    as_json.to_json(*)
  end
end
